### ISSUE TITLE
[Question] About cron.php: Does it really work?

### ISSUE THREAD

**Issue by wagnered** [ampache/ampache#2382]
Could someone please explain how this works: Maybe I'm missing a critical piece of information.

Tbe following is what I understand:

The script is run using php.  The modern webserver runs either php-cgi or php-fpm which I believe manages a pool of php-cgi processes. Php-fpm runs continuously in it's **own process space** while the webserver is running.

As I understand the Linux system, the cron script run by php is  a separate process from  the php-fpm, or php-cgi process created by the webserver. To get this to work correctly, the cron script needs to be run once by Ampache upon its startup. That way, it will be in the same process space created by the webserver and can fill the cache in that process space.

Even if the cron.php file was executed using php-fpm outside the webserver process, the system would create a separate process and it would be isolated from the webserver's php-fpm process.

Of course this begs the question: is a cron task necessary?  I say "no".

BTW & FYI & FWIW: I just learned that if the webservers are running php-fpm, init.php would need to be run only once and not in every top-level script because of php-fpm's persistence.

**Comment by lachlan-00** [comment]
you're right, fpm is loading processes all the time so parts of this won't work.

It would need to move into memcached to be effective this way

**Comment by wagnered** [comment]
You might consider [Redis](https://redis.io/documentation) which is BSD licensed.

Memcached is meant to be, but doesn't need to be a distributed(cloud) in-memory cache system.  It does one thing and it does it well.

Redis can be an in-memory cache, and database with on-disk persistence, and a publish/subscriber messaging system.  Thought I'd suggest it because of it's apparent versatility.

I didn't take a close look at the data types of each system, but I I remember Ampache caching only basic data types.

**Comment by lachlan-00** [comment]
i'm using redis for nextcloud and it seems simple enough. i'll look at that first

**Comment by lachlan-00** [comment]
it seems to be that a lot of these solutions are good for storing strings related to values but not really for storing arrays of data? 

redis and memcached don't seems to have the same drop in requirements of the live php arrays/indexes

**Comment by wagnered** [comment]
Whew...  I thought I was going senile.  I took another look at cron.inc and the live cache statements were gone.  I **eventually** decided to check the history and discovered why. :smile:

What I haven't discovered is the purpose of compute_cache.inc.  when I looked at its original commit, I didn't see any corresponding changes to Ampache : Specifically, changes to access table "cache_object_count".  Were the changes made in some unrelated commit?  If so, that's a bad habit to get into.

Back to the limitations of redis and memcached.  I found a blog by [Dmitry Polyakovsky](http://dmitrypol.github.io/redis/2017/06/21/redis-complex-data-struct.html) titled "Storing complex data structures in Redis". It explains how to emulate an array with list or hash data types.  Don't know how well it would adapt to the needs of Ampache, though.

Could you give me an idea as to under what circumstances filling the cache with the contents of the database would be advantageous?  Maybe updating the catalogs? I believe there are enough negatives to rethink this, especially with an unlimited cache.

I would like to give thought to reinventing the existing ampache cache.  For example, perhaps php could be built with zts thread safety making it multithreading.  A worker thread could act as a cache handler.  

Also, the cleaning, adding, and updating of the catalogs would probably benefit Ampache if threads are spawned to do those tasks.

**Comment by lachlan-00** [comment]
the compute cache loads up all the queries like counts for plays based on users to allow large sites to handle the load a bit better. dogmazic and afterster created that and it's actually a pretty decent way of doing things.

I almost want to merge smart lists into something like that too but it could blow up too much for a database. I'll check the multithreading and the blog article out as well. something different does need to be done.

The easy answer of why is that when i cache ahead i can load the albums page in 0.2 seconds. without caching it takes about 4.5s.

There was a CVE for phpmailer dropped a few days ago so i'm going to change pace to release that fix to master instead of looking at features.

**Comment by wagnered** [comment]
Concerning the cache, how about a file based cache?  Not quite as fast as in-memory, but for a server with a huge user base, it might be a good compromise: not necessary to worry about memory limitations.

I've done more research on multithreading and it has its own logistic problems.  First, I haven't found a distribution that provides a thread safe version. 

I found a site or two that does provide a pre-built thread safe php package, but I don't know if they provide all the thread safe modules that Ampache requires. [Windows.php.net](https://windows.php.net/download/) provides both php  .exes for Windows.
